File: src/pointer_ring/pointer_ring.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <signal.h>

/*

gcc -O3 pointer_ring.c -o pointer_ring_c -lrt
gcc -O3 -g3 -ggdb pointer_ring.c -o pointer_ring -lrt

./pointer_ring 4096000 99999999
./pointer_ring 4294967296 99999999

*/

int main(int argc, char **argv) {
    size_t pointer_count = 0;
    size_t buffersize;
    size_t iterations;
    if (argc >= 3) {
        buffersize = atoll(argv[1]);
        iterations = atoll(argv[2]);
    } else {
        /* buffersize = 5 * 1024 * 1024; */
        return 1;
    }

    // allocate buffer
    size_t *buffer = malloc(buffersize * sizeof(size_t));
    printf("buffer start: %p\n", buffer);
    printf("buffer   end: %p\n", buffer + buffersize);

    puts("initializing with pointers to next...");
    // fill with pointers-to-next
    for (size_t i=0; i<buffersize; i++) {
        buffer[i] = i+1;
    }
    // set last one to point to first
    buffer[buffersize-1] = 0;

    puts("randomizing buffer...");
    // permute buffer
    for (size_t i=0; i<10*buffersize; i++) {
        size_t idx1 = rand() % buffersize;
        size_t idx2 = rand() % buffersize;

        // start with:
        // a -> b -> c
        // d -> e -> f
        size_t a = buffer[idx1];
        size_t b = buffer[a];
        size_t c = buffer[b];
        size_t d = buffer[idx2];
        size_t e = buffer[d];
        size_t f = buffer[e];

        // end with:
        // a -> e -> c
        // d -> b -> f
        buffer[a] = e;
        buffer[d] = b;
        buffer[b] = f;
        buffer[e] = c;
    }

    struct timespec start;
    struct timespec end;
    memset(&start, 0, sizeof(struct timespec));
    memset(&end, 0, sizeof(struct timespec));
    clock_gettime(CLOCK_MONOTONIC, &start);


    // follow pointer chain
    puts("starting pointer traversal..");
    size_t idx = buffer[0];
    for (size_t i=0; i<iterations; i++) {
        idx = buffer[idx];
        /* printf("value: %zx\n", idx); */
    }
    // print final value to prevent it from being optimized out entirely
    printf("final value: %zx\n", idx);

    clock_gettime(CLOCK_MONOTONIC, &end);
    double nanoseconds = end.tv_nsec - start.tv_nsec;
    nanoseconds += 1e9 * (1.0*end.tv_sec - 1.0*start.tv_sec);
    printf("nanoseconds: %f\n", nanoseconds);

    return 0;
}

File: src/pointer_ring/pointer_ring.rs
/*
rustc pointer_ring.rs -O -o pointer_ring_rust
*/
#![feature(rand)]

use std::env;
extern crate rand;
use std::__rand::thread_rng;
use rand::distributions::{IndependentSample, Range};
// use std::time::precise_time_ns;
use std::time::Instant;

fn main() {
    for argument in env::args() {
        println!("{}", argument);
    }

    let args: Vec<String> = env::args().collect();
    let buffersize: usize = args[1].parse().unwrap();
    let iterations: usize = args[2].parse().unwrap();

    let mut buffer: Vec<usize> = (1..(buffersize+1)).into_iter().collect();
    // make last go to first
    buffer[buffersize-1] = 0;

    let range = Range::new(0, buffersize);
    let mut rng = thread_rng();

    println!("randomizing buffer");
    for _ in 0..(10*buffersize) {
        let idx1 = range.ind_sample(&mut rng);
        let idx2 = range.ind_sample(&mut rng);

        // start with:
        // a -> b -> c
        // d -> e -> f
        let a = buffer[idx1];
        let b = buffer[a];
        let c = buffer[b];
        let d = buffer[idx2];
        let e = buffer[d];
        let f = buffer[e];

        // end with:
        // a -> e -> c
        // d -> b -> f
        buffer[a] = e;
        buffer[d] = b;
        buffer[b] = f;
        buffer[e] = c;
    }

    let start = Instant::now();

    let mut idx = buffer[0];
    for _ in 0..iterations {
        idx = buffer[idx];
        // println!("value: {}", idx);
    }
    println!("final value: {}", idx);

    let end = Instant::now();
    let duration = end.duration_since(start);
    let nanoseconds = (duration.subsec_nanos() as f64) + 1e9 * (duration.as_secs() as f64);
    println!("time: {}", nanoseconds);
}
